[api-minor] Downsize large image masks before rendering

apoorv-mishra opened the pull request:
Hi,

This PR addresses #8076 . The problem was the large image mask resolution. The solution was to scale it down in `image.js` and then send the reduced image mask data for rendering via `evaluator.js`.

For downscaling, I used 'point sampling' algorithm which does something like [this](https://www.imagemagick.org/Usage/filter/#point). Since we're dealing with image mask(b&w palette), this would serve the purpose well. Compare [this](https://i.imgur.com/KrTFwR3.png)(pdfium) vs [this](https://i.imgur.com/ICGlJr2.png).

Thanks,
Apoorv

apoorv-mishra committed:
Downsize image mask, fixes #8076

Code cleaning

Early return if resize factors equal one

Restored blame and added new page render param

Fixed indentation and JSDoc

Fix linting errors

Fix style issues

Don't downsize small image masks

Fix linter errors

Downsize a dimension only if it crosses certain threshold value

Tweak mininum threshold for downsizing image masks

Downsize not more than 4096x4096

Snuffleupagus commented on src/display/api.js:
```diff
@@ -820,12 +820,22 @@ var PDFPageProxy = (function PDFPageProxyClosure() {
         enable: !getDefaultSetting('disableWebGL'),
       });
 
+      var combinedInitialTransform;
```
While it might be OK to have this new resizing behaviour be the default, it must be possible for an API consumer to opt-out of it (since it's a new feature).
Hence I'd suggest adding a new render parameter (with accompanying JSDoc), for example named  `resizeLargeImages`, which defaults to `true` but can be used to disable resizing.

So something like the following:
```javascript
let combinedInitialTransform = null;
if (params.resizeLargeImages !== false) {
  if (params.transform) {
    combinedInitialTransform = Util.transform(params.transform,
                                              params.viewport.transform);
  } else {
    combinedInitialTransform = params.viewport.transform;
  }
}
```

Snuffleupagus commented on src/core/image.js:
```diff
@@ -75,6 +75,74 @@ var PDFImage = (function PDFImageClosure() {
+    var resizedHeight = destHeight;
+    if (resizeFactorX === 1 && resizeFactorY === 1) {
+      dest = src;
+      return { dest, resizedWidth, resizedHeight, };
```
If you're passing the `src` through unchanged, I'd assume that you need to also return the *original* width/height as well to prevent rendering errors!?

Snuffleupagus commented on src/core/image.js:
```diff
@@ -75,6 +75,74 @@ var PDFImage = (function PDFImageClosure() {
+    var dest = new Uint8Array(destWidth * destHeight);
+    var resizedWidth = destWidth << 3;
+    var resizedHeight = destHeight;
+    if (resizeFactorX === 1 && resizeFactorY === 1) {
```
Another option could perhaps be to simply return `null` in this case, e.g.
```javascript
function downsizeImageMask(src, w1, h1, w2, h2) {
  let resizeFactorX = Math.round(w1 / w2);
  let resizeFactorY = Math.round(h1 / h2);
  if (resizeFactorX === 1 && resizeFactorY === 1) {
     return null;
  }
  
  // The rest of the code here...
}
```
And then change the [call-site](https://github.com/mozilla/pdf.js/pull/9255/files#diff-7105ab0e122d5ff915e2f8e96347d136R344) to do something like this instead:
```javascript
  let resizedImageObj = downsizeImageMask(data, width, height,
                                          resizeWidth, resizeHeight);
  if (resizedImageObj) {
    return {
      data: resizedImageObj.dest,
      width: resizedImageObj.resizedWidth,
      height: resizedImageObj.resizedHeight,
    };
  }
}
```

Snuffleupagus reviewed (commented):
Currently, this patch only seem to resize large image *masks*, but *normal* images can and often are much larger than necessary too. Don't we need to resize regular images as well?

I've not looked in detail at the resizing algorithm implemented in the patch, however I've left a few inline comments regarding the general structure.
Please also keep the following in mind: When adding new code, please use `let` rather than `var`; and please avoid unnecessarily changing the blame on lines where it's not necessary.

apoorv-mishra commented:
Hi @Snuffleupagus , you are indeed right. We need to resize regular images too. But my main motive here was to fix the issue at hand. Also, navigating through the codebase, I clearly saw sort of 'separation of concerns' in handling regular images and image masks(one example being, `createImageMask` and `createImageData`). So, this patch cleanly fixes the issue without breaking other things. Also, there would not be a problem later to handle regular images(due to this separation of concerns). Though, in that case, we would be needing more sophisticated algorithms, something like(I guess) interpolation coupled with box sampling for which I would be needing some time to study. I would be highly obliged to extend this to handle regular images, if I get the time to do so. I would be pushing the changes suggested by you shortly.

Snuffleupagus commented:
> We need to resize regular images too. But my main motive here was to fix the issue at hand.

Thanks, I just wanted to ask since this was something I noticed while looking through the code.
Unless @yurydelendik thinks otherwise, that can probably be implemented in a follow-up instead.

Snuffleupagus commented on src/core/image.js:
```diff
@@ -75,6 +75,73 @@ var PDFImage = (function PDFImageClosure() {
+      }
+      // Add padding
+      if (mask) {
+          while (mask) {
```
Nit: Please remove two spaces of indentation on this line.

Snuffleupagus commented on src/display/api.js:
```diff
@@ -713,6 +713,9 @@ var PDFDocumentProxy = (function PDFDocumentProxyClosure() {
  *                    radial gradient) or a CanvasPattern object (a repetitive
  *                    image). The default value is 'rgb(255,255,255)'.
+ * @property {boolean} resizeLargeImages - (optional) Whether or not large
+ *                     images need to be resized before rendering. The default
```
Replace `need to` with `will` here, since that seems more accurate.

yurydelendik commented on src/core/image.js:
```diff
@@ -75,6 +75,73 @@ var PDFImage = (function PDFImageClosure() {
+          mask = 128;
+          val = 0;
+        }
+        let elemMask = 128;
```
Rename `elemMask` to `srcMask`

yurydelendik commented on src/core/image.js:
```diff
@@ -75,6 +75,73 @@ var PDFImage = (function PDFImageClosure() {
+          elemMask >>= bitHop;
+        }
+        if (!mask) {
+          dest[destIndex++] = val;
```
Move this to the beginning of the loop if (at line 118).

yurydelendik commented on src/core/image.js:
```diff
@@ -75,6 +75,73 @@ var PDFImage = (function PDFImageClosure() {
+      }
+      // Add padding
+      if (mask) {
+        while (mask) {
```
replace the loop with `val += mask | (mask - 1)`;

yurydelendik commented on src/core/evaluator.js:
```diff
@@ -382,10 +385,20 @@ var PartialEvaluator = (function PartialEvaluatorClosure() {
         var imgArray = image.getBytes(bitStrideLength * height);
         var decode = dict.getArray('Decode', 'D');
 
+        let resizeWidth = null, resizeHeight = null;
```
default `resizeWidth` to `width` and `resizeHeight` to `height`

yurydelendik commented on src/core/image.js:
```diff
@@ -271,6 +339,19 @@ var PDFImage = (function PDFImageClosure() {
       }
     }
 
+    if ((resizeWidth > 0 && resizeWidth < width) &&
```
we will require the resizeWidth and resizeHeight to have a value, see comment above. let's use `resizeWidth < width || resizeHeight < height`to decide if we need to downsize.

yurydelendik reviewed (approved):
Looks good with changes below. I did not analyze the performance or quality of the algorithm -- we can improve that later.

brendandahl commented:
@apoorv-mishra looks like this PR is close, can you address the above comments?

apoorv-mishra commented:
@brendandahl So sorry, due to other commitments, I haven't been able to look at this. You're correct, this is close. I'll take a closer look at this during the coming weekend.

redlum94 commented:
@apoorv-mishra Do you already have any news on this PR?

apoorv-mishra commented:
Hi @brendandahl and @yurydelendik , apologies for keeping this PR pending for so long. I have addressed the above comments pointed out by @yurydelendik and resubmitted the patch. Please have a look.

timvandermeij commented:
/botio test

Snuffleupagus commented on src/core/image.js:
```diff
@@ -79,6 +79,68 @@ var PDFImage = (function PDFImageClosure() {
+    }
+    let destWidth = (Math.ceil(packedWidth / xHop) * (8 / bitHop) + 7) >> 3;
+    let destHeight = Math.ceil(h1 / yHop);
+    let dest = new Uint8Array(destWidth * destHeight);
```
This must be `Uint8ClampedArray` instead, since all image handling code has been refactored to return data in that format. (If you rebase the patch against `master`, you'll most likely get even more failures unless the typed array is fixed here.)

Snuffleupagus commented on src/core/image.js:
```diff
@@ -79,6 +79,68 @@ var PDFImage = (function PDFImageClosure() {
+    }
+    let packedWidth = (w1 + 7) >> 3;
+    let xHop = (resizeFactorX >> 3);
+    xHop = xHop ? xHop : 1;
```
Can't you simply fold the two lines above together as `let xHop = (resizeFactorX >> 3) || 1;`, rather than first initializing it and then using a ternary condition?

Snuffleupagus commented on src/core/image.js:
```diff
@@ -79,6 +79,68 @@ var PDFImage = (function PDFImageClosure() {
+    let xHop = (resizeFactorX >> 3);
+    xHop = xHop ? xHop : 1;
+    let yHop = resizeFactorY;
+    let bitHop;
```
Nit: Could initialize this to `1` and remove the `else` case below.

Snuffleupagus reviewed (commented):
Looking at the test-results this seems to, among other things, break Type3 fonts quite badly. One possible, somewhat hacky, solution for the font related regressions might be to add `type3Evaluator.combinedInitialTransform = null;` just after line https://github.com/mozilla/pdf.js/blob/master/src/core/evaluator.js#L2654 to fix that.

apoorv-mishra commented:
@Snuffleupagus , Are you sure that's the font issue? Those are image masks as far as I can tell by debugging. Can you be specific about the pdf file? I debugged with the file named fit11-talk.pdf and it turned out the problem was with image masks. I didn't look at the other files though. Anyways, I'll try making the change suggested by you above and see what happens.

yurydelendik commented:
- [ ] we need to sort out caching before landing it -- change in combined transforms invalidates cached image

timvandermeij commented:
@apoorv-mishra Do you perhaps have time to finish this? I see that @yurydelendik posted some final remarks above and then we should be able to merge this.

timvandermeij commented on src/core/evaluator.js:
```diff
@@ -378,10 +381,22 @@ var PartialEvaluator = (function PartialEvaluatorClosure() {
 
+        let resizeWidth = width, resizeHeight = height;
+        if (this.combinedInitialTransform) {
+          let combinedTransform =
```
This can be `const` instead of `let`.

apoorv-mishra replied:
Done.

timvandermeij commented on src/core/evaluator.js:
```diff
@@ -378,10 +381,22 @@ var PartialEvaluator = (function PartialEvaluatorClosure() {
+        if (this.combinedInitialTransform) {
+          let combinedTransform =
+            Util.transform(this.combinedInitialTransform, ctm);
+          let scaledDimensions =
```
This can be `const` instead of `let`.

apoorv-mishra replied:
Done.

timvandermeij commented on src/core/image.js:
```diff
@@ -260,12 +319,14 @@ var PDFImage = (function PDFImageClosure() {
   PDFImage.createMask = function({ imgArray, width, height,
-                                   imageIsFromDecodeStream, inverseDecode, }) {
+                                   imageIsFromDecodeStream, inverseDecode,
+                                   resizeHeight, resizeWidth, }) {
```
It looks a bit better to reverse width and height here, just like in the rest of the patch, i.e., first width, then height.

apoorv-mishra replied:
Done.

timvandermeij commented on src/core/image.js:
```diff
@@ -301,6 +362,18 @@ var PDFImage = (function PDFImageClosure() {
     }
 
+    if (resizeWidth < width || resizeHeight < height) {
+      let resizedImageObj = downsizeImageMask(data, width, height,
```
This can be `const` instead of `let`.

apoorv-mishra replied:
Done.

Snuffleupagus commented:
>     * [ ]  we need to sort out caching before landing it -- change in combined transforms invalidates cached image

This part of the puzzle probably isn't very "beginner friendly", given all of the cases to consider and that caching occurs on both worker *and* main threads. I'll attempt, time permitting of course, to look into this during the weekend.

apoorv-mishra commented:
We also need to decide on the "min" threshold value of `width` and `height` for which to downsize the image mask, like don't downsize `if (width < 1000 && height < 1000)`. Some of the tests are failing because we haven't decided on this `MIN_THRESHOLD` check.

apoorv-mishra commented:
Any progess on this @Snuffleupagus ?

Snuffleupagus commented:
> Any progess on this

Yes, I've unfortunately been side-tracked by other issues, but now I'm slowly getting there :-)

Fixing the  `imageCache` wasn't as straightforward as I'd hoped and my initial attempt ended up requiring way too much re-factoring. I'm now trying a slightly less invasive approach; hopefully I'll have a patch ready later in the week.

timvandermeij commented:
The reference tests show that many image masks are downsized too much and therefore become blurry. This should be fixed first before figuring out how to do caching, because that's of no use if the functionality is not complete yet. Could you check why the reference images are regressing? You can generate the reference images locally for testing; see https://github.com/mozilla/pdf.js/wiki/Contributing.

apoorv-mishra commented:
> We also need to decide on the "min" threshold value of `width` and `height` for which to downsize the image mask, like don't downsize `if (width < 1000 && height < 1000)`. Some of the tests are failing because we haven't decided on this `MIN_THRESHOLD` check.

@timvandermeij This is what I suggest. What do you think?

timvandermeij commented:
I think a resize threshold of 1000 should be fine. If it's a constant we can always tweak it. I would say that a value that makes most reference tests look good should be OK, but note that I'm not too familiar with the image decoding/resizing code to say for sure.

apoorv-mishra commented:
Hello @timvandermeij , can we run ref tests on these changes? I can't run locally, browser abruptly stops when I try to run `gulp makeref`(checked out on latest master). I keep getting something like `Exiting due to channel error`

Snuffleupagus commented:
> can we run ref tests on these changes?

That's really of no use at this point, since the patch is still trivially broken.
The added dimension checks will only prevent issues with square images, but does absolutely nothing for the *much* more common case of (non-square) rectangular images. For example consider an image that is `100 x 1100`, in which cause you'd obviously only want to resize *one* of the dimensions.

apoorv-mishra commented:
Ah, sorry for being sloppy @Snuffleupagus. Honestly speaking, I thought about it, but missed it while submitting the patch. Made the corrections now. Unfortunately, still can't get around to run refs locally, so directly submitted here.

timvandermeij commented:
/botio test

apoorv-mishra commented:
> For example consider an image that is 100 x 1100, in which case you'd obviously only want to resize one of the dimensions.
As I doubted, resizing one dimension doesn't seem to be a good idea @Snuffleupagus . Some of the refs fail due to that.

Snuffleupagus commented:
> Unfortunately, still can't get around to run refs locally, so directly submitted here.

Unless you actually run tests locally, a patch like this which touches code in the `core/` folder is basically impossible to work on effectively (as you're no double aware of at this point in time).

> As I doubted, resizing one dimension doesn't seem to be a good idea

How could you reasonably have doubted that, when you didn't run any tests locally!?
Besides, just because that didn't immediately work doesn't mean that the idea itself is incorrect, since the latest update will still perform *really* poorly on extremely narrow and large images. (You may simply have "lucked out" on choosing a limit which masks the problem for now, but it doesn't solve anything as-is.)

There's also the question if the downsizing algorithm is even correct and/or good, but I've got neither the time/interest to look into this further (given the time already spent).

apoorv-mishra commented:
> Unless you actually run tests locally, a patch like this which touches code in the core/ folder is basically impossible to work on effectively (as you're no double aware of at this point in time).

I understand. But, you freshly clone the repo, you run `gulp makeref`, you successfully generate refs. Then you run `gulp test`(yes friend, on `master` synced) and you see `OHNOES! Some tests failed` . I can't understand why that happens. Quite deceptive.

> As I doubted, resizing one dimension doesn't seem to be a good idea 
May be I was wrong about it considering the fact that this would somehow screw up the aspect ratio. But it doesn't work this way. I strongly feel that algorithm works correctly(even for resizing just one dimension) and does what it's supposed to do. My original version was overhauled by @yurydelendik himself. So you may consider the algo coming from him more than me. The only problem was that it "aggressively" downsizes masks. Now, that's a problem because some images are so sensitive that resizing their dimensions even by a factor of 0.75 shows perceptible crooked edges. So, I forced it to downsize by not more than 4096x4096.

> but I've got neither the time/interest to look into this further (given the time already spent).
I understand that your time is important. I am a relatively modest programmer compared to many others working on pdfjs and the likes of you. But this kind of statement(especially the "interest" part) discourages new contributors. Even if I leave the issue, other new contributors would probably hesitate to pick it up, if we write such statements.

Snuffleupagus commented:
Closing for now, since there's not been any updates for a couple of years and as-is this patch cannot land given that there's too many regressions in existing test-cases; issue #8076 remains open to track this.
